### PR TITLE
Let PETSc  make the decision on PCSide

### LINKED ISSUE #8681: Strange convergence behavior if using CG
### Description of the enhancement or error report
I have a SPD system and want to take advantage by using the CG solver. However, the solve is noticeable slower than with GMRES and the convergence shows a linear residual of 0, but continues to iterate. Exodiff confirms that both GMRES and CG give the same result, however CG takes many times the number of iterations.

### Rationale for the enhancement or information for reproducing the error

Modifying the example `test/tests/kernels/2d_diffusion/2d_diffusion_test.i` to

```
[Mesh]
  file = square.e
[]

[Variables]
  active = 'u'

  [./u]
    order = FIRST
    family = LAGRANGE
  [../]
[]

[Kernels]
  active = 'diff'

  [./diff]
    type = Diffusion
    variable = u
  [../]
[]

[BCs]
  active = 'left right'

  [./left]
    type = DirichletBC
    variable = u
    boundary = 1
    value = 0
  [../]

  [./right]
    type = DirichletBC
    variable = u
    boundary = 2
    value = 1
  [../]
[]

[Executioner]
  type = Steady

  solve_type = 'PJFNK'

  petsc_options_iname = '-pc_type -pc_hypre_type -ksp_type'
  petsc_options_value = 'hypre boomeramg cg'

[]

[Outputs]
  file_base = out
  exodus = true
[]
```

CG convergence:
```
 0 Nonlinear |R| = 1.732051e+00
      0 Linear |R| = 0.000000e+00
      1 Linear |R| = 0.000000e+00
      2 Linear |R| = 0.000000e+00
      3 Linear |R| = 0.000000e+00
      4 Linear |R| = 0.000000e+00
      5 Linear |R| = 0.000000e+00
      6 Linear |R| = 0.000000e+00
 1 Nonlinear |R| = 1.769225e-08
      0 Linear |R| = 0.000000e+00
      1 Linear |R| = 0.000000e+00
      2 Linear |R| = 0.000000e+00
      3 Linear |R| = 0.000000e+00
      4 Linear |R| = 0.000000e+00
      5 Linear |R| = 0.000000e+00
      6 Linear |R| = 0.000000e+00
      7 Linear |R| = 0.000000e+00
      8 Linear |R| = 0.000000e+00
      9 Linear |R| = 0.000000e+00
     10 Linear |R| = 0.000000e+00
 2 Nonlinear |R| = 0.000000e+00
```
GMRES convergence:
```
 0 Nonlinear |R| = 1.732051e+00
      0 Linear |R| = 1.732051e+00
      1 Linear |R| = 4.836823e-02
      2 Linear |R| = 3.331048e-16
 1 Nonlinear |R| = 1.055803e-08
```

### PR DESCRIPTION
**Pull request by fdkong**
Different solvers in PETSc come with different default settings.
The right preconditioning is invalid for some solvers in PETSc, e.g., CG.

It is better to have no default value in MOOSE.

Refs #8681

### CODE REVIEW

**Comment by permcody** [comment]
Hmm, Tagging @friedmud

**Comment by fdkong** [comment]
Logic is still not right. `NonlinearSystemBase::setPCSide` always get called.

**Comment by YaqiWang** [comment]
Need to add a test like what Hans did. Also this will change the behavior when we do not have Preconditioning input block because nowhere calls `NonlinearSystemBase::setPCSide`, right?

**Comment by andrsd** [comment]
I am not sure this _closes_ #8681. I'd rather say it refs it, since the problem there is that Dirichlet-type BC is producing an non-symmetrical matrix and therefore CG fails, correct?

**Comment by fdkong** [comment]
> Need to add a test like what Hans did. Also this will change the behavior when we do not have Preconditioning input block because nowhere calls NonlinearSystemBase::setPCSide, right?

`No Preconditioning bock `and  `PCSide not set ` have the same the behavior because users do not change pcsideflag.

File: framework/src/preconditioners/MoosePreconditioner.C [hunk 1]
```diff
@@ -27,7 +27,7 @@ InputParameters validParams<MoosePreconditioner>()
   InputParameters params = validParams<MooseObject>();
   params.addPrivateParam<FEProblemBase *>("_fe_problem_base");
 
-  MooseEnum pc_side("left right symmetric", "right");
+  MooseEnum pc_side("left right symmetric");
```
Comment (andrsd) [inline framework/src/preconditioners/MoosePreconditioner.C]: What do you think about adding `default` value which would be the default ;-) And if it set, we do not change the preconditioner side...

**Comment by fdkong** [comment]
> I am not sure this closes #8681. I'd rather say it refs it, since the problem there is that Dirichlet-type BC is producing an non-symmetrical matrix and therefore CG fails, correct?

In his case, CG fails just because the right preconditioning is used, although he also has the issue that the matrix is not symmetric.  PETSc CG supports the left preconditioning only.

**Comment by andrsd** [comment]
> PETSc CG supports the left preconditioning only.

So with left preconditioning, the matrix does not have to be symmetrical?

File: framework/src/preconditioners/MoosePreconditioner.C (hunk 1 above)
Comment (fdkong) [inline framework/src/preconditioners/MoosePreconditioner.C]: how about `none`? In PETSc we have `left right symmetric` but not `none`. If `none` is set, we let PETSc make the decision.

Comment (permcody) [inline framework/src/preconditioners/MoosePreconditioner.C]: `default` is better than `none`. The former actually means use whatever PETSc thinks, while the latter implies "don't do anything".

**Comment by fdkong** [comment]
> So with left preconditioning, the matrix does not have to be symmetrical?

CG always requires the matrix symmetrical. The test case posted by @hr87 is small and trivial, and it therefore still can be solved by CG  even though the matrix is not symmetrical. 

It is a good idea  not to  close this issue. Because they may run into issues when matrix becomes large.

**Comment by andrsd** [comment]
> CG always requires the matrix symmetrical. The test case posted by @hr87 is small and trivial, and it therefore still can be solved by CG even though the matrix is not symmetrical.

Ok, that's what I thought...

File: framework/src/preconditioners/MoosePreconditioner.C (hunk 1 above)
Comment (fdkong) [inline framework/src/preconditioners/MoosePreconditioner.C]: OK, good idea. Let's use `default`.

**Comment by andrsd** [comment]
So, if using CG on a non-symmetrical matrix is technically wrong, why are we doing that in the test?

**Comment by fdkong** [comment]
> So, if using CG on a non-symmetrical matrix is technically wrong, why are we doing that in the test?

Changed penalty boundary conditions.

**Comment by fdkong** [comment]
All failed tests are due to `EXODIFF`. There are two solutions:

(1) Regold the tests 

(2) or add a petsc option `-ksp_pc_side right` to all failed tests.

Any suggestion? @permcody @andrsd @YaqiWang

**Comment by permcody** [comment]
This change is currently too disruptive to have any hope of going in. Here are two ways we can introduce this change:

1. Only allow PETSc to pick the default for CG or perhaps **not** GMRES if that works. This should fix the immediate issue without causing tons of test diffs that you will have to go through.

2. Since I do believe that this is a superior design, it might be time to introduce a new "legacy" option in MOOSE. We've done this before and have gotten pretty good at it. All apps will start with legacy behavior turned on, over time we'll fix individual tests, then whole apps then eventually default to legacy behavior being turned off.

**Comment by fdkong** [comment]
> This change is currently too disruptive to have any hope of going in. Here are two ways we can introduce this change:
> 
>     Only allow PETSc to pick the default for CG or perhaps not GMRES if that works. This should fix the immediate issue without causing tons of test diffs that you will have to go through.
> 
>     Since I do believe that this is a superior design, it might be time to introduce a new "legacy" option in MOOSE. We've done this before and have gotten pretty good at it. All apps will start with legacy behavior turned on, over time we'll fix individual tests, then whole apps then eventually default to legacy behavior being turned off.

Let me try the first one because it is trivial. If the first one does not work, I will try the second one.

File: framework/src/utils/PetscSupport.C [hunk 2]
```diff
@@ -459,15 +459,20 @@ petscSetDefaults(FEProblemBase & problem)
   SNES snes = petsc_solver->snes();
   KSP ksp;
   SNESGetKSP(snes, &ksp);
-
+  PetscBool  isgmres;
+  PetscObjectTypeCompare((PetscObject)ksp,KSPGMRES,&isgmres);
```
Comment (andrsd) [inline framework/src/utils/PetscSupport.C]: Spacing here is all wrong.

**Review by andrsd — DISMISSED** [review]
Let's refactor the PC side stuff into its own function, it is getting big.

**Comment by YaqiWang** [comment]
It is not too bad in Rattlesnake, only `yak:ThermalRadiation.MarshakWaveBenchmark_test` needs further investigation, for others we can just regold. Are we still using right preconditioning for GMRES by default?

**Comment by fdkong** [comment]
> It is not too bad in Rattlesnake, only yak:ThermalRadiation.MarshakWaveBenchmark_test needs further investigation, for others we can just regold. Are we still using right preconditioning for GMRES by default?

Yes, I have to. Otherwise there are many failed tests.

**Comment by andrsd** [comment]
> Are we still using right preconditioning for GMRES by default?

I believe it was the default from the beginning of MOOSE. And that's why the default in `FEProblem` was set to `RIGHT`. It would destroy the whole MOOSE world... ;-)

**Comment by YaqiWang** [comment]
I think we are exclusively using GMRES, if we are not changing the preconditioning, why does the diff come from?

**Comment by fdkong** [comment]
> I think we are exclusively using GMRES, if we are not changing the preconditioning, why does the diff come from?

Because the `default` in MOOSE is different from the `default` in PETSc.

**Comment by permcody** [comment]
Changing from left to right changes the numerics quite a bit. It can change the converged value and all sorts of things.

**Comment by YaqiWang** [comment]
But you told me we are using right preconditioning for GMRES by default in this change.

**Comment by friedmud** [comment]
Do not change the default from being right preconditioning.  Right
preconditioning gives much better convergence analysis for our whacky
multiphysics systems.

Other than that, go nuts with allowing people to specify it... or even
trying to detect when someone is using CG and switch to "left"
automatically.  But don't undo something that every single run of MOOSE
since the beginning of time has been using just because someone decided to
use CG on a problem that it couldn't even be used for...

**Comment by permcody** [comment]
> Do not change the default from being right preconditioning.  Right
preconditioning gives much better convergence analysis for our whacky
multiphysics systems.

Are you sure? This is a rather absolute blanket statement. The more people that use the framework, the more we find these kinds of global framework choices getting in the way. This change caused several dozen tests to fail but it looks like the majority of them aren't convergence problems. More like convergence differences. I like the idea of giving PETSc more power to do what it thinks is best but I also don't want to cause pain for people. We could ping Jed or somebody else for guidance on this.

**Comment by hr87** [comment]
@friedmud The diffusion operator is self-adjoint and the resulting FEM matrix is supposed to be SPD, a highly desirable property in linear algebra. So the decision to use a solver that is highly effective for an SPD matrix on a problem that is supposed to be SPD should not be surprising. The diffusion operator is the prime example where CG CAN be used. What is way more surprising is that Moose destroys a SPD matrix. Yes, GMRES works fine on a small example problem, but given larger problems with several thousand variables, using CG to avoid the necessary storage of 50+ solutions vectors makes good sense.

**Comment by friedmud** [comment]
@permcody I am sure.  There has to be a default choice.  For us, right is a better choice for a default.

I don't have any problem detecting a choice of CG and switching to left though.  And we really should provide an easy (input file) way of selecting the side.

@hr87 MOOSE never makes any attempt to maintain an SPD matrix.  As you've found out, even just using DirichletBCs means you end up with a non-SPD matrix, but there are a million other ways MOOSE destroys SPD matrices as well.  MOOSE is meant to solve problems that are more difficult than pure diffusion... and those problems are almost never SPD.

It's a choice.  MOOSE will not always be the most optimal for specific types of problems... but it will work well.

The tradeoff in not being optimal for specific (simple) problems is outweighed by having one single code path that is simple and straightforward.  If we worried about maintaining SPD matrices we would have tremendous amounts of code everywhere just to fiddle with the matrix... and that code has its own runtime penalty associated with it (which would be unnecessary in any problem which is not SPD... which are most of the problems being solved by MOOSE).

**Comment by friedmud** [comment]
@permcody one more thing: the choice of side won't matter much on tests.  They are too small and don't typically show the major issues with the choice of PC side.

**Comment by fdkong** [comment]
> Do not change the default from being right preconditioning. Right preconditioning gives much better convergence analysis for our whacky multiphysics systems. 

It is hard for me to imagine  that the side of preconditioner has a BIG impact on the converge. Do you have an example on this? I am curious to explore what is the actual reason causing this issue.  

>Other than that, go nuts with allowing people to specify it... or even trying to detect when someone is using CG and switch to "left" automatically. But don't undo something that every single run of MOOSE since the beginning of time has been using just because someone decided to use CG on a problem that it couldn't even be used for...

There are some solvers that may support either  the `left` or the `right` side only. For example, `CG` supports `left` only, and `fGMRES` supports `right` only. The null-space problems have to be solved with `left` preconditioners. We could let `PETSc` help  our users make this kind of choices.  Otherwise, users may have no idea what they should choose.

BTW,  `trying to detect when someone is using CG ` is not easy in MOOSE because at the time we set petsc options, the solver is not setup not. In other words, it is too early to detect which solvers are chosen by  users.

**Comment by fdkong** [comment]
In PETSc, we use `left` preconditioning as default, which is opposite to MOOSE. http://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/KSP/KSPSetPCSide.html

**Comment by permcody** [comment]
Yeah, I'd like to get @jedbrown's thoughts on this:

MOOSE has always set the preconditioning side to the right. However, Fande is proposing that we toss this job back to the solver by default. Users will be able to override this as needed but it seems like a reasonable framework choice. @friedmud suggests otherwise. 

He's right about one thing, changing this default will be difficult with the history we've built forcing to just one side but we've looked at the testing and it is doable if we do it over time. Are there advantages to making this choice for users like this? Should we let the solver pick?

Thanks

**Comment by jedbrown** [comment]
Do you really care which "side" the PC is applied on, or do you care about which norm the KSP works in? If you care about the norm, call `KSPSetNormType(ksp,KSP_NORM_UNPRECONDITIONED)`. I think the "side" is a relatively uninteresting anachronism unless you're specifically interested in, for example, whether a block-triangular preconditioner yields two repeated eigenvalues with a minimal polynomial of degree 2 or whether it produces distinct eigenvalues with a minimal polynomial of degree 2.

**Comment by fdkong** [comment]
> Do you really care which "side" the PC is applied on, or do you care about which norm the KSP works in? If you care about the norm, call KSPSetNormType(ksp,KSP_NORM_UNPRECONDITIONED). 

Thanks,  @jedbrown.

Does the type of norm affect the converge?  The right preconditioning is usually paired with `KSP_NORM_UNPRECONDITIONED`, and `KSP_NORM_PRECONDITIONED` is chosen if the left side is used. 

We care about the `pc side`  because we know some KSP, such as CG,  does not work with the `right pc `(which is the default choice in MOOSE).

>I think the "side" is a relatively uninteresting anachronism unless you're specifically interested in, for example, whether a block-triangular preconditioner yields two repeated eigenvalues with a minimal polynomial of degree 2 or whether it produces distinct eigenvalues with a minimal polynomial of degree 2.

@friedmud  found that the `right preconditioning` works much better  than the` left preconditioning` for multiphysics problems, which is the reason that he set the` right pc ` as the default choice a long time ago.

Is this because the right preconditioning uses `KSP_NORM_UNPRECONDITIONED`?  `KSP_NORM_UNPRECONDITIONED` works better than `KSP_NORM_PRECONDITIONED`? Do you have any thoughts on this?

**Comment by jedbrown** [comment]
Of course the norm affects convergence. It changes the space in which you look for a solution and (usually more significantly) it changes the declaration of convergence. I think most people choose RIGHT because they don't trust their preconditioner and thus want UNPRECONDITIONED. Penalty boundary conditions basically require the PRECONDITIONED norm because otherwise the residual drops by the penalty (often `1e10`) on the first iteration merely because the boundary conditions become satisfied. But if your preconditioner is nearly singular (e.g., BoomerAMG applied directly to Taylor-Hood Stokes) then you misdiagnose stagnation for convergence.

**Comment by fdkong** [comment]
> Of course the norm affects convergence. It changes the space in which you look for a solution and (usually more significantly) it changes the declaration of convergence. 

But does not mean that `UNPRECONDITIONED` is always better than `PRECONDITIONED`?  It is a problem-dependent thing.

>I think most people choose RIGHT because they don't trust their preconditioner and thus want UNPRECONDITIONED.

This makes sense a lot!  For real applications, it is nontrivial   to construct good preconditioenrs.  But,  `UNPRECONDITIONED` should also work with `LEFT` from my understanding. If so, we should choose the `norm type` instead of `pc side`.

> Penalty boundary conditions basically require the PRECONDITIONED norm because otherwise the residual drops by the penalty (often 1e10) on the first iteration merely because the boundary conditions become satisfied. But if your preconditioner is nearly singular (e.g., BoomerAMG applied directly to Taylor-Hood Stokes) then you misdiagnose stagnation for convergence.

Is there any way to detect the property of the preconditioner, and choose a correct norm (`PRECONDITIONED`, `UNPRECONDITIONED` or `natural`)?  Any suggestions?

**Comment by jedbrown** [comment]
My recommendation is to *only* set the norm. The KSP will automatically choose the side that makes sense for that implementation.

**Comment by fdkong** [comment]
> My recommendation is to only set the norm. The KSP will automatically choose the side that makes sense for that implementation.

Thanks, @jedbrown.  It should be chosen as  `UNPRECONDITIONED` in MOOSE, I think.

**Comment by friedmud** [comment]
Sorry I'm late here: Jed (as usual) has this right. I chose right
preconditioning to get at the unpreconditioned norm. That is the important
part. If we can select the unpreconditioned norm in spite of the side...
then let's do that.

I found many years ago that for the types of problems we solve the
unpreconditioned norm gave MUCH better convergence analysis.

**Comment by fdkong** [comment]
Add ` KSPSetNormType()` as the preferred approach to specify the solver.  We will eventually remove the `pcside` from moose, (not in this PR).  There are just a few cases which use `pcside`, and for those cases we can just use petsc options directly.

**Comment by friedmud** [comment]
This seems ok to me.

**Comment by permcody** [comment]
@fdkong - This can go in when you get a chance to rebase it.

**Comment by permcody** [comment]
Fande, Please rebase and we can merge this.

**Comment by fdkong** [comment]
> Fande, Please rebase and we can merge this.

@permcody, I re-based the code.  There are some fails in `zapdos`.

**Comment by fdkong** [comment]
Ping, @permcody

**Review by permcody — APPROVED** [review]

**Comment by lindsayad** [comment]
Ran across this beautiful history while @roystgnr and I were talking about https://github.com/idaholab/moose/pull/27446 today

**Pull request merged**